keps/sig-auth/20190408-node-scoped-daemonset.md

```diff
@@ -0,0 +1,239 @@
+---
+title: Node-Scoped DaemonSet
+authors:
+  - "@haiyanmeng"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-auth
+  - sig-apps
+reviewers:
+  - "@liggitt"
+  - "@tallclair"
+  - "@mikedanese"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-04-08
+last-updated: 2019-04-08
+status: provisional
+---
+
+# Node-Scoped DaemonSet
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+* [Background](#background)
+    * [NodeAuthorizer and NodeRestriction](#nodeauthorizer-and-noderestriction)
+    * [Per-Pod Identity of DaemonSets](#per-pod-identity-of-daemonsets)
+* [Definition](#definition)
+* [Proposal](#proposal)
+    * [The Origin of Node-Scoped Info](#the-origin-of-node-scoped-info)
+    * [Workflow of NodeAuthorizer for DaemonSet Requests](#workflow-of-nodeauthorizer-for-daemonset-requests)
+    * [Workflow of NodeRestriction for DaemonSet Requests](#workflow-of-noderestriction-for-daemonset-requests)
+    * [Implementation Details](#implementation-details)
+* [Design Details](#design-details)
+    * [Test Plan](#test-plan)
+    * [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+Provide a mechanism restricting a
+[DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/)
+pod to only managing the resources that reside on the same node as the DaemonSet
+pod.
+
+## Motivation
+
+Currently, a DaemonSet pod can manage resources in a whole cluster. When one
+DaemonSet pod is compromised, the whole cluster will be exposed to attack.
+
+## Background
+
+### NodeAuthorizer and NodeRestriction
+
+Currently,
+[NodeAuthorizer](https://kubernetes.io/docs/reference/access-authn-authz/node/)
+and
+[NodeRestriction](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction)
+can be used to limit a kubelet's API access. In order to be authorized by
+NodeAuthorizer and limited by the NodeRestriction admission controller, kubelets
+must use a credential that identifies them as being in the system:nodes group,
+with a username of system:node:\<nodeName\>.
+
+### Per-Pod Identity of DaemonSets
+
+Before v1.10, a single service account was used to run all the pods of a
+DaemonSet, and APIServer cannot identify requests from different pods of a
+DaemonSet.
+
+With the introduction of
+[BoundServiceAccountToken](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/bound-service-account-tokens.md)
```

Comment:
BoundServiceAccountTokens are set up by the Kubelet, and the credential is obtained through a token request. Is there already logic in the NodeRestriction controller to prevent the Kubelet identity from requesting a token bound to a pod on another node?

Reply:
Yes. NodeRestriction already covers this:
https://github.com/kubernetes/kubernetes/blob/3ec638d3cd75794dc7b034389b5065f7139b8570/plugin/pkg/admission/noderestriction/admission.go#L142